### PR TITLE
feat(CRT-352): create RegistrationService CRD

### PR DESCRIPTION
**Pull request by MatousJobanek**
## Description
It will be used as a holder of reg-service configuration as well as a triggerrer of the actual deployment.
The params that are in the CRD have been taken from the ConfigMap used by the registration service

## Checks
1. Have you run `make generate` target? **[yes/no]**

**yes**

2. Does `make generate` change anything in other projects (host-operator, member-operator, toolchain-common)? **[yes/no]** 
(NOTE: You can ignore it if the only changes are in the annotation `alm-examples` of the `ClusterServiceVersion` object)

**yes**

3. In case other projects are changed, please provides PR links.
    - host-operator: https://github.com/codeready-toolchain/host-operator/pull/108

### CODE REVIEW

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (alexeykazakov) [inline pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: We can delete this comment.

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+// RegistrationServiceSpec defines the desired state of RegistrationService
+// +k8s:openapi-gen=true
+type RegistrationServiceSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
+
+	// The image identifies which image of the registration service should be used for a deployment
+	Image string `json:"version"`
```
Comment (alexeykazakov) [inline pkg/apis/toolchain/v1alpha1/registrationservice_types.go]:
```suggestion
	Image string `json:"image"`
```

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+	// The number of replicas of the deployed registration service
+	Replicas int `json:"replicas,omitempty"`
```
Comment (alexeykazakov) [inline pkg/apis/toolchain/v1alpha1/registrationservice_types.go]:
```suggestion
	// +optional
	Replicas int `json:"replicas,omitempty"`
```

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+	// The environment identifies which mode the registration service should be running in - prod, stage, e2e-tests, dev, etc.
+	Environment string `json:"environment,omitempty"`
```
Comment (alexeykazakov) [inline pkg/apis/toolchain/v1alpha1/registrationservice_types.go]:
```suggestion
	// +optional
	Environment string `json:"environment,omitempty"`
```

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+	// The AuthClient contains all necessary information about the auth client
+	AuthClient AuthClient `json:"authClient,omitempty"`
```
Comment (alexeykazakov) [inline pkg/apis/toolchain/v1alpha1/registrationservice_types.go]:
```suggestion
	// +optional
	AuthClient AuthClient `json:"authClient,omitempty"`
```

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+type AuthClient struct {
+	// The LibraryUrl identifies the auth library location
+	LibraryUrl string `json:"libraryUrl,omitempty"`
+
+	// The Config contains the auth config
+	Config string `json:"config,omitempty"`
+
+	// The PublicKeysUrl identifies the public keys location
+	PublicKeysUrl string `json:"publicKeysUrl,omitempty"`
```
Comment (alexeykazakov) [inline pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: All these fields are not optional, are they? If so then let’s remove omitempty

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+// RegistrationServiceStatus defines the observed state of RegistrationService
+// +k8s:openapi-gen=true
+type RegistrationServiceStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
+
+	// Conditions is an array of current Registration Service deployment conditions
+	// Supported condition types:
+	// Deploying, and Deployed
```
Comment (alexeykazakov) [inline pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: Wouldn’t just one condition Ready with multiple reasons: Deploying, Deployed, Failing, etc, be enough and easier to handle?

Comment (alexeykazakov) [reply pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: Just for clarification. Let’s keep conditions as an array but IMO for now let’s use only Ready condition with multiple reasons.

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go (hunk 6 above)
Comment (MatousJobanek) [reply pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: They are optional - at least they are right now in the registration-service

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go (hunk 7 above)
Comment (MatousJobanek) [reply pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: yes, I wrote it in the wrong way - I wanted to write: "Supported condition reasons" :-)

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go (hunk 1 above)
Comment (MatousJobanek) [reply pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: stupid me - I forgot to push a newer commit with updated/fixed stuff

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go (hunk 6 above)
Comment (MatousJobanek) [reply pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: ok, now I probably understand what you mean by that. In the current registation-service, there are the default values used, so I guess that I should move the default values from registation service into the host-operator and fill it into the CR

Comment (MatousJobanek) [reply pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: on the other hand, wouldn't it make more sense to keep these default values inside the registration-service as they logically belong there?

**Comment by MatousJobanek** [comment]
@alexeykazakov thanks for your review - I accidentally forgot to push the latest commit that changed most of the problematic parts. Anyway, I merged your suggestions and fixed the missing ones + added additional columns and it should be done now.

File: pkg/apis/toolchain/v1alpha1/registrationservice_types.go (hunk 6 above)
Comment (alexeykazakov) [reply pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: Let’s keep the default values inside the reg service.
And yes the AuhClient should be optional in CRD but if it’s specified that it doesn’t make much sense to keep individual fields of the AuthClient optional. Does it make sense to you?

Comment (MatousJobanek) [reply pkg/apis/toolchain/v1alpha1/registrationservice_types.go]: ok, changed in 5da71b3

**Comment by MatousJobanek** [comment]
> I would keep the fields in AuthClient required (while the AuthClient itself being optional) but it’s minor.

changed in 5da71b3 

> Also we could add print column for Environment too but not sure how useful it would be...

add both image and environment cda551b 

Thanks a lot for your review :+1:

**Pull request merged**